packages/notebook/src/widget.ts

```diff
@@ -1047,19 +1056,61 @@ export class StaticNotebook extends WindowedList<NotebookViewModel> {
         this.cellsArray.forEach((cell, i) => {
           const estHeight = this._viewModel.estimateWidgetSize(i);
           cell.node.style.containIntrinsicSize = `auto ${estHeight}px`;
+          (cell.node.style as any).contentVisibility = 'auto';
+          cell.node.style.contain = 'layout style paint';
         });
       });
 
-      // Watch for newly added cells and set intrinsic size for them too
-      this.model?.cells.changed.connect(() => {
-        requestAnimationFrame(() => {
-          this.cellsArray.forEach((cell, i) => {
-            const estHeight = this._viewModel.estimateWidgetSize(i);
-            cell.node.style.containIntrinsicSize = `auto ${estHeight}px`;
-          });
+      // Use the real browser viewport for visibility detection
+      this._setupContentVisibilityObserver();
+
+      // Disconnect observer when the notebook is detached
+      const disconnectOnDetach = () => {
+        if (this._contentVisibilityObserver) {
+          this._contentVisibilityObserver.disconnect();
+          this._contentVisibilityObserver = null;
+        }
+        this.node.removeEventListener('lm:before-detach', disconnectOnDetach);
+      };
+      this.node.addEventListener('lm:before-detach', disconnectOnDetach);
```

Comment:
Should we handle this in

https://github.com/jupyterlab/jupyterlab/blob/77f02be2ebb8a7154223ead742d4cd6eee5ec6a6/packages/notebook/src/widget.ts#L2392-L2394 

instead?

I did not know lumino emits events like that.

Reply:
Oh, good catch! If that was intentional, I would expect there to be a comment with explanation, so I think this is a mistake. It means we are not calling some logic wrt windowing in:

https://github.com/jupyterlab/jupyterlab/blob/c832df73b105c9f3fc215b8aec1180c8805e9c12/packages/ui-components/src/components/windowedlist.ts#L1160-L1166